src/frontend/src/helpers/Websocket/WebsocketClient.ts

```diff
@@ -205,6 +211,92 @@ class SeashellWebsocket {
     }
   }
 
+  // Connects the socket, sets up the disconnection monitor
+  public async connect(): Promise<void> {
+    // Failure - probably want to prompt the user to attempt to reconnect or
+    //  log in again
+    this.onFailure(async () => {
+      clearInterval(this.timeoutInterval);
+      try {
+        await this.invoke_cb('failed');
+      } catch (err) {
+        console.error(err);
+      }
+    });
+
+    // Socket closed - probably want to prompt the user to reconnect
+    this.onClose(async () => {
+      clearInterval(this.timeoutInterval);
+      try {
+        await this.invoke_cb('disconnected');
+      } catch (err) {
+        console.error(err);
+      }
+    });
+
+    if(!this.authenticated) {
+      await this.invoke_cb('failed');
+      throw new WebsocketError("socket is not authenticated");
+    }
+
+    this.debug && console.log("Seashell socket set up properly");
+    this.timeoutInterval = setInterval(async () => {
+      try {
+        if (this.timeoutCount++ === 3) {
+          this.invoke_cb('timeout');
+        }
+        await this.ping();
+        if (this.timeoutCount >= 3) {
+          this.invoke_cb('timein');
+        }
+        this.timeoutCount = 0;
+      } catch (err) {
+        console.error(err);
+      }
+    }, 4000);
+    this.requests[-3].callback = this.io_cb;
+    this.requests[-4].callback = this.test_cb;
+    this.debug && console.log("Websocket disconnection monitor set up properly.");
+    // Run the callbacks.
+    await this.invoke_cb('connected');
+  }
+
+  public disconnect(): void {
+    this.websocket.close();
+  }
+
+   public register_callback(type: string, cb: (message?: any) => any, now?: boolean) : number {
+    this.callbacks[this.key] = new Callback(type, cb, now);
+
+    if(type === 'disconnected' && !this.isConnected() && now) {
+      cb();
+    } else if(type === 'connected' && this.isConnected() && now) {
+      cb();
+    } else if(type === 'failed' && this.failed && now) {
+      cb();
+    }
+    return this.key++;
+  }
+
```

Comment:
For the event handling I think it's probably better to find some library to emit events that react can listen on.
I don't know how react wants it. Will need to ask Kevin.